CountingBloomFilter.hpp

```diff
@@ -271,61 +252,76 @@ template<typename T>
 	}
 	readHeader(file);
-	struct stat buf;
-	if (fstat(fileno(file), &buf) != 0) {
-		std::cerr << "ERROR: Failed to open file: " << path << "\n";
-		exit(EXIT_FAILURE);
-	}
-	size_t arraySizeOnDisk = buf.st_size - sizeof(struct FileHeader);
-	if (arraySizeOnDisk != m_sizeInBytes) {
-		std::cerr << "ERROR: File size of " << path << " (" << arraySizeOnDisk << " bytes), "
-		          << "does not match size read from its header (" << m_sizeInBytes << " bytes).\n";
-		exit(EXIT_FAILURE);
-	}
-
-	size_t nread = fread(m_filter, arraySizeOnDisk, 1, file);
-	if (nread != 1 && fclose(file) != 0) {
+	char* filter;
+	filter = new char[m_sizeInBytes];
+	file.read(filter, m_sizeInBytes);
+	m_filter = reinterpret_cast<T*>(filter);
+	if (!file) {
 		std::cerr << "ERROR: The byte array could not be read from the file: " << path << "\n";
 		exit(EXIT_FAILURE);
 	}
+	file.close();
 }
 
 template<typename T>
 void
-CountingBloomFilter<T>::readHeader(FILE* file)
+CountingBloomFilter<T>::readHeader(std::istream& file)
 {
-	FileHeader header;
-	if (fread(&header, sizeof(struct FileHeader), 1, file) != 1) {
-		std::cerr << "Failed to read header\n";
-		exit(EXIT_FAILURE);
-	}
-	if (header.hlen != sizeof(FileHeader)) {
-		std::cerr << "Bloom Filter header length: " << header.hlen
-		          << " does not match expected length: " << sizeof(FileHeader)
-		          << " (likely version mismatch)" << std::endl;
-		exit(EXIT_FAILURE);
-	}
-	if (memcmp(header.magic, MAGIC_HEADER_STRING, MAGIC_LENGTH) != 0) {
-		std::cerr << "Bloom Filter type does not match" << std::endl;
+	std::string magic_header(MAGIC_HEADER_STRING);
+	(magic_header.insert(0, "[")).append("]");
+	std::string line;
+	std::getline(file, line);
+	if (line.compare(magic_header) != 0) {
+		std::cerr << "ERROR: magic string does not match (likely version mismatch) \n";
+		std::cerr << "Your magic string:                " << line << "\n";
+		std::cerr << "CountingBloomFilter magic string: " << magic_header << std::endl;
 		exit(EXIT_FAILURE);
 	}
-	if (header.version != BloomFilter_VERSION) {
-		std::cerr << "Bloom Filter version does not match: " << header.version
-		          << " expected: " << BloomFilter_VERSION << std::endl;
-		exit(EXIT_FAILURE);
+
+	/* Read bloom filter line by line until it sees "[HeaderEnd]"
+	   which is used to mark the end of the header section and
+	   assigns the header to a char array*/
+	std::string headerEnd = "[HeaderEnd]";
+	char* toml_buffer = new char[0];
+	while (std::getline(file, line)) {
+		if (line == headerEnd) {
+			int currPos = file.tellg();
+			if (toml_buffer != nullptr) {
+				delete[] toml_buffer;
+			}
+			toml_buffer = new char[currPos];
+			file.seekg(0, file.beg);
+			file.read(toml_buffer, currPos);
+			file.seekg(currPos, file.beg);
+			break;
+		}
 	}
-	m_size = header.size;
-	m_hashNum = header.nhash;
-	m_kmerSize = header.kmer;
+
+	// Send the char array to a stringstream for the cpptoml parser to parse
+	std::stringstream toml_stream;
```

Comment:
```suggestion
	std::istringstream toml_stream(toml_buffer);
```
Use `istringstream` or `ostringstream`. Avoid `stringstream`.